packages/type-utils/tests/TypeOrValueSpecifier.test.ts

```diff
@@ -0,0 +1,288 @@
+      [{ from: 'package', name: 'MyType', package: 42 }],
+      [{ from: [], name: 'MyType' }],
+      [{ from: ['file'], name: 'MyType' }],
+      [{ from: ['lib'], name: 'MyType' }],
+      [{ from: ['package'], name: 'MyType' }],
+      [
+        {
+          from: 'package',
+          name: ['MyType', 'myValue'],
+          package: ['./filename.js', './another-file.js'],
+        },
+      ],
+      [
+        {
+          from: 'package',
+          name: 'MyType',
+          path: './filename.js',
+          unrelatedProperty: '',
+        },
+      ],
+    ])("doesn't match a malformed package specifier: %s", runTestNegative);
+  });
+
+  describe('typeMatchesSpecifier', () => {
+    function runTests(
+      code: string,
+      specifier: TypeOrValueSpecifier,
+      expected: boolean,
+    ): void {
+      const rootDir = path.join(__dirname, 'fixtures');
+      const { ast, services } = parseForESLint(code, {
+        project: './tsconfig.json',
+        filePath: path.join(rootDir, 'file.ts'),
+        tsconfigRootDir: rootDir,
+      });
+      const type = services
+        .program!.getTypeChecker()
+        .getTypeAtLocation(
+          services.esTreeNodeToTSNodeMap.get(
+            (ast.body[0] as TSESTree.TSTypeAliasDeclaration).id,
+          ),
+        );
+      expect(typeMatchesSpecifier(type, specifier, services.program!)).toBe(
+        expected,
+      );
+    }
+
+    function runTestPositive(
+      code: string,
+      specifier: TypeOrValueSpecifier,
+    ): void {
+      runTests(code, specifier, true);
+    }
+
+    function runTestNegative(
+      code: string,
+      specifier: TypeOrValueSpecifier,
+    ): void {
+      runTests(code, specifier, false);
+    }
+
+    it.each<[string, TypeOrValueSpecifier]>([
+      ['interface Foo {prop: string}; type Test = Foo;', 'Foo'],
+      ['type Test = RegExp;', 'RegExp'],
+    ])('matches a matching universal string specifier', runTestPositive);
+
+    it.each<[string, TypeOrValueSpecifier]>([
+      ['interface Foo {prop: string}; type Test = Foo;', 'Bar'],
+      ['interface Foo {prop: string}; type Test = Foo;', 'RegExp'],
+      ['type Test = RegExp;', 'Foo'],
+      ['type Test = RegExp;', 'BigInt'],
+    ])(
+      "doesn't match a mismatched universal string specifier",
+      runTestNegative,
+    );
+
+    it.each<[string, TypeOrValueSpecifier]>([
+      [
+        'interface Foo {prop: string}; type Test = Foo;',
+        { from: 'file', name: 'Foo' },
+      ],
+      [
+        'interface Foo {prop: string}; type Test = Foo;',
+        { from: 'file', name: ['Foo', 'Bar'] },
+      ],
+      [
+        'interface Foo {prop: string}; type Test = Foo;',
+        { from: 'file', name: 'Foo', path: 'tests/fixtures/file.ts' },
+      ],
+      [
+        'interface Foo {prop: string}; type Test = Foo;',
+        {
+          from: 'file',
+          name: ['Foo', 'Bar'],
+          path: 'tests/fixtures/file.ts',
+        },
+      ],
```

Comment:
🤔 looking at the test failures locally for this one, in the `typeDeclaredInFile` function:

* `relativePath` is `'tests/fixtures/file.ts'`
* `absolutePath` is `'/Users/josh/repos/typescript-eslint/packages/type-utils/tests/fixtures/file.ts`
* `declarationFiles[0].fileName` is `'/users/josh/repos/typescript-eslint/packages/type-utils/tests/fixtures/file.ts'`

@marekdedic my instinct is to `.toLowerCase()` them for now, then file a followup issue to investigate whether this can cause edge case bugs / there's a better way. Thoughts?

Reply:
@marekdedic I really want to get this in - so I'll apply this change now and file the followup issue. Cheers!

Reply:
https://github.com/typescript-eslint/typescript-eslint/issues/6697

Reply:
Ha, sorry, missed the comment, I'll take a look